instrumentation/action_pack/lib/opentelemetry/instrumentation/action_pack/patches/action_controller/metal.rb

```diff
@@ -14,9 +14,15 @@ module Metal
             def dispatch(name, request, response)
               rack_span = OpenTelemetry::Instrumentation::Rack.current_span
               if rack_span.recording?
-                rack_span.name = "#{self.class.name}##{name}" unless request.env['action_dispatch.exception']
+                route = rails_route(request)
```

Comment:
One blocker here, we should only calculate the rails route either the span naming is configured to use the normalized path, or if the http.route `enable_recognize_route` configuration is enabled.

The reason being as it is written right now if neither of those options are configured we are still paying the cost of calculating route even if it does not get used anywhere.

I'm aware that this is nitpicky, but the concern here is that we want to minimize the cost associated with instrumenting applications. Most code gets slow slowly through many changes of this nature. However I do think this is a valuable change as it introduces consistency and compliance with the semantic conventions. 

We just want to make sure we allow users to opt out if they choose.

I think we could memoize this here to get the desired effect, however I don't necessarily expect you to use my suggestion below so long as the code is changed to support avoiding calculating the route when it doesn't get used.
```rb
           def dispatch(name, request, response)
              rack_span = OpenTelemetry::Instrumentation::Rack.current_span
              if rack_span.recording?
                unless request.env['action_dispatch.exception']
                  rack_span.name = case instrumentation_config[:span_naming]
                                   when :controller_action then "#{self.class.name}##{name}"
                                   else "#{request.method} #{rails_route(request)}"
                                   end
                end

                rack_span.set_attribute('http.route', rails_route(request)) if instrumentation_config[:enable_recognize_route]

                rack_span.set_attribute('http.target', request.filtered_path) if request.filtered_path != request.fullpath
              end

              super(name, request, response)
            end

            private

            def rails_route(request)
              @rails_route ||= ::Rails.application.routes.router.recognize(request) do |route, _params|
                return route.path.spec.to_s
                # Rails will match on the first route - see https://guides.rubyonrails.org/routing.html#crud-verbs-and-actions
              end
            end
```

Reply:
@robertlaurin I would like to confirm that `Metal` object instances are reinitialized per request. Introducing instance variables here may result in _incorrect routes being recorded_ if the same object instance is reused for different requests. 

Do you happen to know how `Metal` allocates object instances?

Reply:
I was under the impression that controller lifecycles start and end on a per request basis, as rack is the simplest form of a controller I expected that to hold true for this. However I'm not confident enough in my memory to assert this, we should either confirm this or choose a different approach. I have no preference here, so long as the route look up doesn't occur needlessly.

Reply:
It seems the [Metal class itself](https://github.com/rails/rails/blob/8015c2c2cf5c8718449677570f372ceb01318a32/actionpack/lib/action_controller/metal.rb#L247) is using memoization (with the request and response). So I assume we can use it as well.

I have switched this PR to do so, with an additional test checking that memoization doesn't reuse previous values.